### PR TITLE
Make sure Kafka serde autodetection doesn't override connector configuration

### LINKED ISSUE #24322: Config like `mp.messaging.connector.smallrye-kafka.value.serializer` is ignored
### Describe the bug

The build step `SmallRyeReactiveMessagingKafkaProcessor#defaultChannelConfiguration()` currently ignores checking for connector-level config like `mp.messaging.connector.smallrye-kafka.value.serializer`. As a result this configuration has no effect when added to the `application.properties` file because the generated channel-level `RunTimeConfigurationDefaultBuildItem` defaults take precedence.

See also discussion in Zulip: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/.E2.9C.94.20Kafka.3A.20Apicurio.20for.20dev.20and.20Confluent.20for.20prod.20mode.3F

### Expected behavior

Explicit connector-level config in `application.properties` should take precedence over generated channel-level defaults.

The build step should probably check for the presence of such properties at build time to determine whether to generate a runtime default or not for a given channel's properties.

#### Use case

We are using Kafka with a schema registry and Avro encoded messages. In production the schema registry is the Confluent implementation and for development in dev mode we use Apicurio. This means that the serializer and deserializer implementation classes are different for production and non-production (i.e. test and development) launch modes. Further, since our application uses multiple channels, we would like to declare the Avro serializer and deserializer at the connector level, rather than individually for each channel. Doing it for each channel results in a lot of duplicated config.

### Actual behavior

The explicit connector-level config (like `mp.messaging.connector.smallrye-kafka.value.serializer`) in the `application.properties` file is ignored, because the autodetection generates corresponding channel-level config which takes precedence.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.7.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### PR DESCRIPTION
**Pull request by Ladicek**
The SmallRye Reactive Messaging Kafka extension automatically detects
key and value serializer/deserializer and sets default config values
for the respective channel. That overrides connector configuration,
which is unexpected. If the user provides connector configuration
themselves, we should not override that. Unfortunately the extension
can only detect connector configuration if it's present during build,
but that will be the case 99.9% of time.

With this commit, if connector configuration is present for key/value
serde, the respective channel configuration is not emitted.

Fixes #24322

### CODE REVIEW

**Review by ozangunalp — APPROVED** [review]

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
This PR was merged too early. It breaks the build. 
It will need to be revisited.

**Comment by cescoffier** [comment]
The revert is: https://github.com/quarkusio/quarkus/pull/24370

**Comment by Ladicek** [comment]
Ugh, let me check.

**Comment by Ladicek** [comment]
OK I know why I didn't see the test failure locally. I did `mvn clean install`, but I should have done `mvn clean install -Dstart-containers` :-/ I'm certainly not a fan of having to always enable Docker-based tests.

**Comment by Ladicek** [comment]
Here's a funny thing: the failing test is `KafkaDevServicesDevModeTestCase`. If I run it in isolation, it passes. If I run `DefaultSerdeConfigTest` before, the test fails. Between these two cases, I can see one difference:

- when I run the test in isolation, autodetection produces `mp.messaging.incoming.prices.value.deserializer = org.apache.kafka.common.serialization.LongDeserializer`
- when I run the tests together, autodetection produces `mp.messaging.incoming.prices.value.deserializer = org.apache.kafka.common.serialization.IntegerDeserializer`

The method annotated `@Incoming("prices")` has the following signature: `double process(int priceInUsd)`. So in fact, the failing case is more correct than the passing case!

All this might be a red herring of course.

**Comment by ozangunalp** [comment]
I noticed the same. And it is not any test run before `KafkaDevServicesDevModeTestCase`, it happens when `DefaultSerdeConfigTest` is run before, even without the added tests. I'll enable more logs on the tested app, maybe it'll tell us something.

**Comment by Ladicek** [comment]
OK, so Reactive Messaging seems to be completely fine. The fact that no consumer connects to Kafka is because the CDI producer method for the `@Channel("processed-prices") Publisher<Double> processedPrices` injection point is never invoked. Hence the publisher never subscribes to its upstream, and that eventually leads to the Kafka consumer not getting subscribed to.

The problem doesn't even seem to be in CDI though. In the failing case, even the `PriceResource.ssePrices` method is never invoked. Need to check how the SSE client in the test works.

**Comment by Ladicek** [comment]
So in the failing case, when the SSE client tries to establish a connection, it gets `javax.ws.rs.ProcessingException: RESTEASY004655: Unable to invoke request: org.apache.http.conn.HttpHostConnectException: Connect to localhost:8081 [localhost/127.0.0.1] failed: Connection refused (Connection refused)` and that exception is just swallowed by RESTEasy (because it happens on an extra thread).

This points to the real problem: in the working case, the `@TestHTTPResource("/prices/stream") URI` correctly points to localhost:8080, but in the failing case, it for some reason points to localhost:8081.

**Comment by ozangunalp** [comment]
My conclusion is `ConfigProvider.getConfig()` has a side effect. On our `getConfig` override, we need to return an empty config.

**Comment by ozangunalp** [comment]
Somehow, the hack in `TestHTTPResourceManager#getUri` doesn't apply when we use `ConfigProvider.getconfig()` in some test.

**Comment by Ladicek** [comment]
That hack is never used anyway, because `TestHTTPConfigSourceProvider` is always present.

One difference I can observe is that in the failing case, `Config.getPropertyNames()` includes `%dev.test.url`, which is not present in the correct case. I don't have an explanation (yet).

**Comment by ozangunalp** [comment]
Oh, you are right about the `TestURLConfigSource`! But I am pretty sure there is a side effect :) When I get rid of `ConfigProvider.getConfig()` in tests everything works.

**Comment by Ladicek** [comment]
For sure. Config is somehow leaking across tests. I wonder when that started happening and why are we seeing it just now :-)

**Comment by Ladicek** [comment]
Hmm so in the correct case, the `ProfileConfigSourceInterceptor` believes the current profile is `dev`, while in the failing case, it thinks the profile is `test`. WAT?

**Comment by Ladicek** [comment]
OK so `io.quarkus.test.junit.BasicLoggingEnabler`, which is a global `BeforeAllCallback` (enabled via `ServiceLoader` file), in its `static` initializer arranges for `BasicLoggingEnabler.buildConfig` to be called on an extra thread (most likely to speed up the `beforeAll` method). The first thing that method does is `ProfileManager.setLaunchMode(LaunchMode.TEST)`.

That happens before the first test (`DefaultSerdeConfigTest`) is even started. The `BasicLoggingEnabler` even creates a `Config` object, but it cleans up after itself. Except the current launch mode remains set. That's probably even intentional.

Then, tests begin. When `DefaultSerdeConfigTest` calls `ConfigProvider.getConfig()`, a global `Config` object is created for the system classloader (which is the TCCL at that point in time, it's a regular JUnit test, so Quarkus doesn't manipulate with TCCL in any way). That global config object is created using `QuarkusConfigFactory`, which explicitly sets a profile to whatever `ProfileManager.getActiveProfile` returns. It returns `test` when current launch mode is `TEST`.

At the beginning of `KafkaDevServicesDevModeTestCase`, the `QuarkusDevModeTest.beforeAll` method calls `ProfileManager.setLaunchMode(LaunchMode.DEVELOPMENT)`, but at that time, the global config object for the system classloader already exists and uses the `test` profile.

To sum up, if there is a sequence of tests that, in the middle, changes the current launch mode, configuration will not work correctly. With the current state of affairs, this can only happen when a `QuarkusDevModeTest` is present in the sequence, because all other kinds of tests (including regular JUnit tests as described above) use the `TEST` launch mode.

**Comment by Ladicek** [comment]
Since this is a more general problem, I have a more general fix. With that, `mvn clean install -Dstart-containers` in the `quarkus-smallrye-reactive-messaging-kafka-deployment` module passes. I'll let CI run in my fork and then submit a PR. When that is incorporated, this PR can be resubmitted without a change.

**Comment by Ladicek** [comment]
FTR, if anyone's interested, here's my attempt to fix the problem: https://github.com/Ladicek/quarkus/commits/launch-mode-in-tests

**Comment by ozangunalp** [comment]
@Ladicek Thanks for the explanation! 
For the fix, I think passing by reflection to avoid creating a config is a bit overkill. 
We can simply get the tccl config from the `ConfigProviderResolver` and release it. It is most likely already created by the `BasicLoggingEnabler`.
I see something similar is done in `TestResourceManager#close`.

**Comment by Ladicek** [comment]
@ozangunalp I want to have a discussion about the change (on a separate PR, which is why I'm not adding the relevant people here) and ideally add an API to SmallRye Config. It's true I could just `ConfigProviderResolver.getConfig(ClassLoader)`, but that forces configuration creation if it doesn't exist yet, which adds unnecessary overhead (even though most of the time, a config _will_ exist). Maybe that's what I end up doing, but I want to have that discussion first.

**Comment by Ladicek** [comment]
Good point about `TestResourceManager` doing that already though. Perhaps I'll change it right away...

**Comment by ozangunalp** [comment]
@Ladicek as you said, it's sane to have the discussion!

**Comment by Ladicek** [comment]
Submitted a draft PR here: 24437

**Comment by Ladicek** [comment]
I'm looking into this again and based on the discussion in #24437, I realized that there's probably an expectation that no `Config` object remains associated with the system classloader. Which indeed the `DefaultSerdeConfigTest` does. Releasing that config object at the end of that test seems to fix things :-)

**Comment by Ladicek** [comment]
Submitted this PR again with a fix to the `DefaultSerdeConfigTest` : #24693
